app/models/manageiq/providers/amazon/cloud_manager/ssa_agent_manager.rb

```diff
@@ -0,0 +1,326 @@
+
+    instance[0].id
+  end
+
+  private
+  def role_exists?(role_name)
+    begin
+      role = iam.role(role_name)
+      role.role_id
+      true
+    rescue ::Aws::IAM::Errors::NoSuchEntity
+      false
+    end
+  end
+
+  def create_role(role_name = MIQ_SSA)
+    return if role_exists?(role_name)
+
+    # Policy Generator:
+    policy_doc = {
+      "Version": "2012-10-17",
+      "Statement": [
+        {
+          "Effect": "Allow",
+          "Principal": {"Service": "ec2.amazonaws.com"},
+          "Action": "sts:AssumeRole"
+        }
+      ]
+    }
+
+    role = iam.create_role({
+      role_name: role_name,
+      assume_role_policy_document: policy_doc.to_json
+    })
+    
+    # grant all priviledges
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonS3FullAccess'
+    })
+    
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonEC2FullAccess'
+    })
+    
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonSQSFullAccess'
+    })
+
+    role
+  end
+
+  def create_profile(profile_name = MIQ_SSA, role_name = MIQ_SSA)
+    ssa_profile = iam.instance_profile(profile_name)
+    ssa_profile = iam.create_instance_profile(instance_profile_name: profile_name) unless ssa_profile.exists?
+
+    create_role(role_name) unless role_exists?(role_name)
+    ssa_profile.add_role(role_name: role_name) if ssa_profile.roles.size == 0
+
+    ssa_profile
+  end
+
+  def get_agent_image_id
+    imgs = ec2.client.describe_images(
+      filters: [
+        {
+          name: "name",
+          values: [ "RHEL-7.3_HVM_GA*" ]
+        }
+      ]
+    ).images
+
+    imgs[0].image_id
+  end
+
+  def create_security_group(group_name = MIQ_SSA)
+    begin
+      sgs = ec2.client.describe_security_groups(
+        filters: [
+          {
+            name: "group-name",
+            values: [ group_name ]
+          }
+        ]
+      ).security_groups
+      return sgs[0].group_id if sgs.length > 0
+
+      # create security group if not exist
+      security_group = ec2.create_security_group({
+        group_name: group_name,
+        description: 'Security group for MIQ SSA Agent',
+        vpc_id: ec2.client.describe_vpcs.vpcs[0].vpc_id
+      })
+
+      security_group.authorize_ingress({
+        ip_permissions: [{
+          ip_protocol: 'tcp',
+          from_port: 22,
+          to_port: 22,
+          ip_ranges: [{
+            cidr_ip: '0.0.0.0/0'
+          }]}]
+      })
+
+      security_group.authorize_ingress({
+        ip_permissions: [{
+          ip_protocol: 'tcp',
+          from_port: 80,
+          to_port: 80,
+          ip_ranges: [{
+            cidr_ip: '0.0.0.0/0'
+          }]}]
+      })
+
+      security_group.authorize_ingress({
+        ip_permissions: [{
+          ip_protocol: 'tcp',
+          from_port: 443,
+          to_port: 443,
+          ip_ranges: [{
+            cidr_ip: '0.0.0.0/0'
+          }]}]
+      })
+
```

Comment:
Instead of repeating these three calls to authorize_ingress explicitly can you set up a loop with a list of the 3 sets of port numbers and make the call inside the loop?  Less code.  Also define the port numbers outside the body of the method -
e.g. PORTS = [22, 80, 443]  ...